pkg/apis/autoscaling/types.go

```diff
@@ -184,6 +197,25 @@ type ResourceMetricSource struct {
 	TargetAverageValue *resource.Quantity
 }
 
+// ExternalMetricSource indicates how to scale on a metric not associated with
+// any Kubernetes object (for example length of queue in cloud
+// messaging service, or QPS from loadbalancer running outside of cluster).
+type ExternalMetricSource struct {
+	// metricName is the name of the metric in question.
+	MetricName string
+	// MetricSelector is used to identify a specific time series
+	// within a given metric.
+	MetricSelector metav1.LabelSelector
+	// TargetValue is the target value of the metric (as a quantity).
```

Comment:
Since these next fields represent a one-of, they should probably be in a struct.  This should document clearly what happens if no value is specified.

e.g

```
external:
  metricName: foobar
  target:
    value: 100
```

What do you think?

Reply:
There is no big difference between grouping 2 fields in a struct and having them "inline". However in other places in this version of the api we have the target inline. So it structs will be kindof inconsistent. We can do it in v2beta2, changing target to structs both in ObjectMetrics and ExernalMetrics.

No value behaviour should obviously be documented.

Reply:
I can accept the consistency argument.  We have very inconsistent use of one-of blocks throughout the system.  Really this should be explicitly discriminated:

```
external:
  metricName: foobar
  target:
    type: Value
    value: 100
```

Reply:
Documented that exactly one of the two must be specified. As @mwielgus wrote - we already have TargetAverageUtilization and TargetAverageValue in resource metrics. I agree with your argument, but I think consistency within a single API object should take precedence in this case. If we used different approach with external and left resource as is we would end up with HPA objects looking like this:
``` 
- resource:
    targetAverageValue:  100
- external:
    target:
      type: Average
      value: 100
```

I believe I'm in agreement with @DirectXMan12 that we want a v2beta2 to (at minimum) add targetAverageValue to object metrics. This would already be backward incompatible, so maybe we can consider changing all targets to structs then?

Reply:
ACK